LanguageFeatures/Patterns/type_inference_A05_t01.dart

```diff
@@ -0,0 +1,36 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion
+/// The context type schema for a pattern p is:
+/// ...
+/// Cast: The context type schema is _.
+///
+/// @description Check that for a cast pattern the context type schema is _
+/// @author sgrekhov22@gmail.com
+
+// SharedOptions=--enable-experiment=patterns,records
+
+import "patterns_lib.dart";
+import "../../Utils/expect.dart";
+
+class A {}
+class B extends A {}
+class C extends B {}
+class D extends C {}
+
+main() {
+  A a = A();
+  B b = B();
+  C c = C();
+  D d = D();
+
+  String log = "";
+  var (v1 as B) = getType(d, (String s) {log += s;});
+  Expect.equals("Object?", log);
+
+  log = "";
+  final (v2 as B) = getType(d, (String s) {log += s;});
+  Expect.equals("Object?", log);
```

Comment:
So how can we test that the type schema is `_` rather than `Object?`?

I'm too tired now. I'll continue tomorrow. ;-)

Reply:
The current analyzer produces an error here. Is this an analyzer issue?
```dart
class A {}
class B extends A {}
class C extends B {}
class D extends C {}

main() {
  A a = A();
  B b = B();
  C c = C();
  D d = D();

  var (v2 as B) = d;
  v2.expectStaticType<Exactly<D>>(); // COMPILE_TIME_ERROR.TYPE_ARGUMENT_NOT_MATCHING_BOUNDS
                                     // 'D Function(D)' doesn't conform to the bound 'B Function(B)' of the type parameter 'R'.
```

Reply:
It looks correct: `v2` gets the declared type `B` because that's the matched value type which is passed on to the subpattern `p` of a cast pattern `p as B`, and there's no error at the initializing expression because the cast pattern has a required type of `Object?`.

Reply:
Then how we can test that the context type schema is `_`? Should, then test llooks like below?
```dart
  var (v1 as B) = b;
  v1.expectStaticType<Exactly<B>>();
  var (v2 as B) = c;
  v2.expectStaticType<Exactly<B>>();
  var (v3 as B) = d;
  v3.expectStaticType<Exactly<D>>();
  ```